### PR TITLE
feat(rollup-relayer): finalize bundle

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

1. add bundle-proposer.
2. finalize bundles.

- [x] db unit tests.
- [x] bundle proposer unit tests.
- [x] l2 relayer unit tests.
- [x] integration tests.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 1]
```diff
@@ -631,6 +663,78 @@ func (r *Layer2Relayer) finalizeBatch(dbBatch *orm.Batch, withProof bool) error
 	return nil
 }
 
+func (r *Layer2Relayer) finalizeBundle(bundle *orm.Bundle, withProof bool) error {
+	dbBatch, err := r.batchOrm.GetBatchByIndex(r.ctx, bundle.EndBatchIndex)
```
Comment (georgehao) [inline rollup/internal/controller/relayer/l2_relayer.go:690]: I think it's better check all the batches of this bundle

Comment (colinlyguo) [reply rollup/internal/controller/relayer/l2_relayer.go:690]: sounds good. added in [3c8c8d2](https://github.com/scroll-tech/scroll/pull/1411/commits/3c8c8d2b77618bed6baced2763d901439d9fa8a1).

File: rollup/mock_bridge/MockBridge.sol
Comment (zimpha) [inline rollup/mock_bridge/MockBridge.sol:1]: can we use [ScrollChainMockFinalize](https://github.com/scroll-tech/scroll-contracts/blob/2ac4f3f7e090d7127db4b13b3627cb3ce2d762bc/src/mocks/ScrollChainMockFinalize.sol) directly  or use script to generate the `MockBridge` from `ScrollChainMockFinalize`?

for exanple, remove `OwnableUpgradeable` and `PausableUpgradeable` related codes (like `OnlySequencer`,  `onlyOwner`, `OnlyProver` and `whenNotPaused`), then flatten `ScrollChainMockFinalize`.

Comment (colinlyguo) [reply rollup/mock_bridge/MockBridge.sol:1]: Yes. We can. could you help me replace this contract (mainly removing verification and l1 messages) with the mock contract in this PR?

**Review by zimpha — APPROVED** [review]

File: coordinator/internal/orm/batch.go [hunk 2]
```diff
@@ -76,7 +76,7 @@ func (*Batch) TableName() string {
 }
 
 // GetUnassignedBatch retrieves unassigned batch based on the specified limit.
-// The returned batch are sorted in ascending order by their index.
+// The returned batches are sorted in ascending order by their indices.
```
Comment (georgehao) [inline coordinator/internal/orm/batch.go]: this `index` is filed, not indices.

File: coordinator/internal/orm/batch.go [hunk 3]
```diff
@@ -93,7 +93,7 @@ func (o *Batch) GetUnassignedBatch(ctx context.Context, startChunkIndex, endChun
 }
 
 // GetAssignedBatch retrieves assigned batch based on the specified limit.
-// The returned batch are sorted in ascending order by their index.
+// The returned batches are sorted in ascending order by their indices.
```
Comment (georgehao) [inline coordinator/internal/orm/batch.go]: ditto

File: coordinator/internal/orm/chunk.go [hunk 4]
```diff
@@ -73,7 +73,7 @@ func (*Chunk) TableName() string {
 }
 
 // GetUnassignedChunk retrieves unassigned chunk based on the specified limit.
-// The returned chunks are sorted in ascending order by their index.
+// The returned chunks are sorted in ascending order by their indices.
```
Comment (georgehao) [inline coordinator/internal/orm/chunk.go]: ditto

File: coordinator/internal/orm/chunk.go [hunk 5]
```diff
@@ -90,7 +90,7 @@ func (o *Chunk) GetUnassignedChunk(ctx context.Context, fromBlockNum, toBlockNum
 }
 
 // GetAssignedChunk retrieves assigned chunk based on the specified limit.
-// The returned chunks are sorted in ascending order by their index.
+// The returned chunks are sorted in ascending order by their indices.
```
Comment (georgehao) [inline coordinator/internal/orm/chunk.go]: ditto

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 6]
```diff
@@ -334,7 +338,7 @@ func (r *Layer2Relayer) ProcessGasPriceOracle() {
 
 // ProcessPendingBatches processes the pending batches by sending commitBatch transactions to layer 1.
 func (r *Layer2Relayer) ProcessPendingBatches() {
-	// get pending batches from database in ascending order by their index.
+	// get pending batches from database in ascending order by their indices.
```
Comment (georgehao) [inline rollup/internal/controller/relayer/l2_relayer.go]: ditto

File: rollup/internal/orm/batch.go [hunk 7]
```diff
@@ -77,7 +81,7 @@ func (*Batch) TableName() string {
 }
 
 // GetBatches retrieves selected batches from the database.
-// The returned batches are sorted in ascending order by their index.
+// The returned batches are sorted in ascending order by their indices.
```
Comment (georgehao) [inline rollup/internal/orm/batch.go]: ditto

File: rollup/internal/orm/batch.go [hunk 8]
```diff
@@ -191,7 +215,7 @@ func (o *Batch) GetRollupStatusByHashList(ctx context.Context, hashes []string)
 }
 
 // GetFailedAndPendingBatches retrieves batches with failed or pending status up to the specified limit.
-// The returned batches are sorted in ascending order by their index.
+// The returned batches are sorted in ascending order by their indices.
```
Comment (georgehao) [inline rollup/internal/orm/batch.go]: pls check all `indices` correctness

File: rollup/internal/config/l2.go [hunk 9]
```diff
@@ -44,3 +44,9 @@ type BatchProposerConfig struct {
 	GasCostIncreaseMultiplier       float64 `json:"gas_cost_increase_multiplier"`
 	MaxUncompressedBatchBytesSize   uint64  `json:"max_uncompressed_batch_bytes_size"`
 }
+
+// BundleProposerConfig loads bundle_proposer configuration items.
```
Comment (amoylan2) [inline rollup/internal/config/l2.go:50]: This struct has not been added to the L2Config struct

Comment (colinlyguo) [reply rollup/internal/config/l2.go:50]: nice catch, added in [d23b617](https://github.com/scroll-tech/scroll/pull/1411/commits/d23b61720a2e8db8ac78191a1d48ea1b18096188).

File: coordinator/internal/orm/batch.go (hunk 2 above)
Comment (colinlyguo) [reply coordinator/internal/orm/batch.go]: got it. all changed to `index fields` in [c96c4dd](https://github.com/scroll-tech/scroll/pull/1411/commits/c96c4dde5822ecb84099d487edbe52cd89299b03).

**Review by georgehao — APPROVED** [review]

**Pull request merged**